cms/src/test/java/uk/nhs/digital/externalstorage/workflow/externalFilePublish/ExternalFilePublishTaskTest.java

```diff
@@ -30,33 +41,163 @@
+
+        executeTask(getDocumentHandleWithVariant(getFinalisedPublication(), UNPUBLISHED), externalFilePublishTask);
+
+        then(s3Connector).should().publishResource(testExternalReference);
+        then(s3Connector).should(never()).unpublishResource(anyString());
+    }
+
+    @Test
+    public void shouldUnpublishUpcomingPublicationResources() throws WorkflowException, RepositoryException {
+        MockJcr.addQueryResultHandler(session, searchResourcesHandle(testExternalReference));
+
+        executeTask(getDocumentHandleWithVariant(getUpcomingPublication(), UNPUBLISHED), externalFilePublishTask);
+
+        then(s3Connector).should(never()).publishResource(anyString());
+        then(s3Connector).should().unpublishResource(testExternalReference);
+    }
+
+    @Test
+    public void shouldUnpublishDatasetResourcesPartOfUpcomingPublication() throws WorkflowException, RepositoryException {
+        MockJcr.addQueryResultHandler(session, searchResourcesHandle(testExternalReference));
+
+        // add Upcoming publication to JCR
+        getDocumentHandleWithVariant(getUpcomingPublication(), UNPUBLISHED);
+
+        executeTask(getDocumentHandleWithVariant(getDataset(), UNPUBLISHED), externalFilePublishTask);
+
+        then(s3Connector).should(never()).publishResource(anyString());
+        then(s3Connector).should().unpublishResource(testExternalReference);
+    }
+
+    @Test
+    public void shouldPublishDatasetResourcesPartOfFinalisedPublication() throws WorkflowException, RepositoryException {
+        MockJcr.addQueryResultHandler(session, searchResourcesHandle(testExternalReference));
+
+        // add Finalised publication to JCR
+        getDocumentHandleWithVariant(getFinalisedPublication(), UNPUBLISHED);
+
+        // when
+        executeTask(getDocumentHandleWithVariant(getDataset(), UNPUBLISHED), externalFilePublishTask);
+
+        then(s3Connector).should().publishResource(testExternalReference);
+        then(s3Connector).should(never()).unpublishResource(anyString());
+    }
+
+    @Test
+    public void shouldPublishFinalisedPublicationAndItsDatasetsResources() throws WorkflowException, RepositoryException {
+        // add Dataset to JCR
+        Node dataset = getDataset();
+        Node publication = getFinalisedPublication();
 
-        executeTask(docNode, externalFilePublishTask);
+        MockJcr.addQueryResultHandler(session, searchResourcesHandle(testExternalReference));
+        MockJcr.addQueryResultHandler(session, searchDatasetsHandle(dataset));
 
-        then(s3Connector).should().publishResource("the/one");
+        // when publishing finalised publication
+        executeTask(getDocumentHandleWithVariant(publication, UNPUBLISHED), externalFilePublishTask);
+
+        // should be executed once for publication and once for dataset
+        then(s3Connector).should(times(2)).publishResource(testExternalReference);
```

Comment:
I think this should use different testExternalReference values for dataset and publication. As it is, the test doesn't prevent a bug where the same reference is used for each of the external resources that need publishing.